### ISSUE TITLE
S4158 FP: Collection filled using delegates

### ISSUE THREAD

**Issue by markusschaber** [SonarSource/sonar-dotnet#4261]
### Description
In one of our unit tests, S4158 triggers false positives, the collection clearly is not empty when S4158 claims it is.

### Repro steps

This is a cut down snippet from one of our unit tests, requires FluentAssertions (or replace with your favorite assertion syntax).
```
namespace StaticAnalyzerS4158Bug
{
    using System;
    using System.Collections.Generic;
    using FluentAssertions;

    static class Program
    {
        static void Main(string[] args)
        {
            var collector = new List<int>();

            DoSomething(collector.Add);

            collector[0].Should().Be(42); // Triggers false positives on S4158.
            collector[1].Should().Be(23); // Triggers false positives on S4158.
        }

        private static void DoSomething(Action<int> callback)
        {
            callback(42);
            callback(23);
        }
    }
}
```

### Expected behavior

The rule does not fire when delegates of any methods which can add members have been passed as callback somewhere.

### Actual behavior

The rule fires 

### Known workarounds

Wrap the delegate in a lambda:
```
            Testee.SubscribeAnalyzerValues(tenant, plcId, i => collector.Add(i));
```

### Related information

* C#/VB.NET Plugins version: SonarLint for Visual Studio 4.33.1.28913
* Visual Studio version: 2019
* MSBuild / dotnet version: .NET Core 3.1
* Operating System: Windows 10

**Comment by costin-zaharia-sonarsource** [comment]
Hi @markusschaber,

thanks for your feedback. I can confirm the false positive.
